packages/eslint-plugin/src/rules/default-param-last.ts

```diff
@@ -24,7 +24,14 @@ export default createRule({
      * @private
      */
     function isOptionalParam(node: TSESTree.Parameter): boolean {
-      return 'optional' in node && node.optional;
```

Comment:
this type of code seems fine since it's not being used for narrowing or type operations. It's just asking... "is this parameter optional". it's basically the same as 
```
@ts-expect-error: optional might not be a key, but that's ok, since we're treating it as false.
return Boolean(node.optional);
```

The trouble with enumerating all the subtypes that can possibly be optional is.... what if you forgot one? And they're unnecessary runtime checks.

To mitigate the exhaustiveness concern, though, you can also do 
```ts
switch (node.type) {
    case AST_NODE_TYPES.ArrayPattern:
    case AST_NODE_TYPES.AssignmentPattern:
    // etc
        return node.optional;
    case // explicitly enumerate ones that _should be excluded_:
    default: 
        const __never_case: never = node.type;
        throw new Error('unexpected type');    
}
```
so you get a compiletime and runtime error if you forget to make the switch exhaustive....

Reply:
I mean, sure, someone _could_ come along and add a type to the union where `optional` really means _not optional_, or it has shape `{ isOptional: false }`, so it's a truthy object.... but I think it would be reasonable to assume that any `TSESTree.Parameter` would only ever have a sensibly defined `.optional` if it has a `.optional` at all.

Reply:
Anyways, the change is fine. I'm just rambling that I don't necessarily think that the change is objectively better _for this specific kind of case_. 🙂

Reply:
Yeah I think the rule is technically speaking a _little_ overly broad... but the things it saves us from are obnoxious enough that I don't mind a little extra flagging. 🤷